ENH: Importing tut. adds multiplexed barcode in sequence section (#454)
qiime2/docs: source/tutorials/importing.rst

With QIIME 2, there are functions to import different types of FASTQ data:

1. FASTQ data with the EMP Protocol format
2. Multiplexed FASTQ data with barcodes in sequences

> 3. FASTQ data with the Casava 1.8 demultiplexed format

Reviewer: Since we are tweaking the list items anyway, how about:

```suggestion
3. FASTQ data in the Casava 1.8 demultiplexed format
```

> 4. Any other kind of FASTQ data

Reviewer: Since we are tweaking the list items anyway, how about:

```suggestion
4. Any other FASTQ data not represented in the list items above
```

> Paired-end "`Earth Microbiome Project (EMP) protocol`_" formatted reads should have three ``fastq.gz`` files total:

Reviewer: Can you also update the Single-end EMP stuff above, similarly, since it too is a directory format, and not a single-file format.

Reviewer: Please apply the same edits you have made in these sections on paired-end EMP to the single-end EMP section, above.

1. one ``forward.fastq.gz`` file that contains the forward sequence reads,
2. one ``reverse.fastq.gz`` file that contains the reverse sequence reads,

> 3. and a third that contains the associated barcode reads

Reviewer: Since we are outlining the other two filenames, probably should explicitly list this one, too:

```suggestion
3. one ``barcodes.fastq.gz`` file that contains the associated barcode reads
```

In this format, sequence data is still multiplexed (i.e. you have only one forward and one reverse ``fastq.gz`` file containing raw data for all of your samples).

> Because you are importing multiple files in a directory, the filenames `forward.fastq.gz` and `reverse.fastq.gz` are *required*.

Reviewer: ```suggestion
Because you are importing multiple files in a directory, the filenames ``forward.fastq.gz``, ``reverse.fastq.gz``, and ``barcodes.fastq.gz`` are *required*.
```

     --input-path emp-paired-end-sequences \
     --output-path emp-paired-end-sequences.qza

.. _`multiplexed barcode in seq import`:

Multiplexed single-end FASTQ with barcode in sequence
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Format description
******************

Users with multiplexed single-ended barcode in sequence reads should have:
1. one ``fastq.gz`` file, containing records from multiple samples,
2. one :doc:`metadata <metadata>` file with a column of per-sample barcodes for use in FASTQ demultiplexing

Obtaining example data
```````````````````````

.. command-block::

   mkdir muxed-se-barcode-in-seqs

.. download::
   :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-se-barcode-in-seqs.fastq.gz
   :saveas: muxed-se-barcode-in-seqs/sequences.fastq.gz

Importing data
**************

.. command-block::

   qiime tools import \
     --type MultiplexedSingleEndBarcodeInSequence \
     --input-path sequences.fastq.gz \
     --output-path multiplexed-seqs.qza

> In this format, sequence data is still multiplexed (i.e. you have only one forward ``fastq.gz`` file containing raw data for all of your samples).

Author: ```suggestion
In this format, sequence data is still multiplexed (i.e. you have only one ``fastq.gz`` file containing raw data for all of your samples).
```

> Because you are importing a single sequence file, your filenames may be whatever you prefer.

Author: ```suggestion
Because you are importing a single file of sequences, your filenames may be whatever you prefer.
```

Each barcode need be represented only once in the metadata mapping file. Barcodes are not required to be in any specific order.

Multiplexed paired-end FASTQ with barcode in sequence
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Format description
******************

Users with multiplexed paired-end barcode in sequence reads should have:
1. one ``forward.fastq.gz`` file, containing forward reads from multiple samples,
2. one ``reverse.fastq.gz`` file, containing reverse reads from the same samples,
3. one :doc:`metadata <metadata>` file with a column of per-sample barcodes for use in FASTQ demultiplexing

In this format, sequence data is still multiplexed (i.e. you have only one forward and one reverse ``fastq.gz`` file containing raw data for all of your samples).

Because you are importing a multi-file directory, the filenames `forward.fastq.gz` and `reverse.fastq.gz` are *required*.

> The order of the records in the ``fastq.gz`` files defines the association between forward and reverse sequence reads, and so a correct order must be preserved. Barcodes in the metadata mapping file are not required to be in any specific order.

Author: ```suggestion
The order of the records in the ``fastq.gz`` files defines the association between forward and reverse sequence reads, so a correct order must be preserved. Barcodes in the metadata mapping file are not required to be in any specific order.
```

Obtaining example data
``````````````````````

.. command-block::

   mkdir muxed-pe-barcode-in-seqs

.. download::

>    :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-pe-barcodes-in-seqs/forward.fastq.gz

Author: ```suggestion
   :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-pe-barcode-in-seqs/forward.fastq.gz
```

   :saveas: muxed-pe-barcode-in-seqs/forward.fastq.gz

.. download::

>    :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-pe-barcodes-in-seqs/reverse.fastq.gz

Author: ```suggestion
   :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-pe-barcode-in-seqs/reverse.fastq.gz
```

     --input-path emp-paired-end-sequences \
     --output-path emp-paired-end-sequences.qza

.. _`multiplexed barcode in seq import`:

> Multiplexed single-end FASTQ with barcode in sequence

Reviewer: Since we technically support dual-index demuxing via the methods in this section, plural barcodes is fine:

```suggestion
Multiplexed single-end FASTQ with barcodes in sequence
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Format description
******************

> Users with multiplexed single-ended barcode in sequence reads should have:

Reviewer: ```suggestion
Users with multiplexed single-ended barcodes in sequence reads should have:
```

1. one ``fastq.gz`` file, containing records from multiple samples,
2. one :doc:`metadata <metadata>` file with a column of per-sample barcodes for use in FASTQ demultiplexing

Obtaining example data
```````````````````````

.. command-block::

   mkdir muxed-se-barcode-in-seq

.. download::
   :url: https://data.qiime2.org/2020.5/tutorials/importing/muxed-se-barcode-in-seq.fastq.gz
   :saveas: muxed-se-barcode-in-seq/sequences.fastq.gz

Importing data
**************

.. command-block::

   qiime tools import \
     --type MultiplexedSingleEndBarcodeInSequence \

>      --input-path sequences.fastq.gz \

Reviewer: This needs to be updated to match your download block above (or, you can alternatively keep this as-is and change the download block).

```suggestion
     --input-path muxed-se-barcode-in-seq/sequences.fastq.gz \
```

     --output-path multiplexed-seqs.qza

In this format, sequence data is still multiplexed (i.e. you have only one ``fastq.gz`` file containing raw data for all of your samples).

Because you are importing a single file of sequences, your filenames may be whatever you prefer.

> Each barcode need be represented only once in the metadata mapping file. Barcodes are not required to be in any specific order.

Reviewer: This is a little confusing upon first read, maybe:

> There should be exactly one unique barcode for each sample you wish to demultiplex. Barcodes are not required to be in any specific order.

Also, we generally don't use the term "mapping" in QIIME 2, that is something we have tried to leave behind in the QIIME 1 world.

Author: I like your language for now. Do you know whether there is an open issue for combinatorial dual indexing, where I can make a note to make this inclusive of unique barcode pairs?

Multiplexed paired-end FASTQ with barcodes in sequence

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Format description
******************

Users with multiplexed paired-end barcode in sequence reads should have:

1. one ``forward.fastq.gz`` file, containing forward reads from multiple samples,
2. one ``reverse.fastq.gz`` file, containing reverse reads from the same samples,

> 3. one :doc:`metadata <metadata>` file with a column of per-sample barcodes for use in FASTQ demultiplexing

Reviewer: ```suggestion
3. one :doc:`metadata <metadata>` file with a column of per-sample barcodes for use in FASTQ demultiplexing (or two columns of dual-index barcodes)
```

"Fastq manifest" formats
~~~~~~~~~~~~~~~~~~~~~~~~

> If your data is not multiplexed (i.e. you have no more than one sample per file) and you don't have either EMP or Casava format, you need to import your data into QIIME 2 manually by first creating a "manifest file" and then using the ``qiime tools import`` command with different specifications than in the EMP or Casava import commands.

Reviewer: ```suggestion
The examples above demonstrate how to import multiplexed data (i.e. EMP or FASTQ files with the barcodes in the sequence) and how to import demultiplexed data that follows some common formatting conventions (the Casava formats). If you don't have data that matches the above cases, you likely will need to import your data into QIIME 2 manually by first creating a "manifest file" and then using the ``qiime tools import`` command with different specifications than in the EMP or Casava import commands.
```
